### PR TITLE
Change building status message

### LINKED ISSUE #1730: Castle screen, creature dwelling, wrong text displayed in info bar at the bottom of the screen when the mouse cursor is positioned over the creature dwelling
In fheroes2, the wrong text is displayed in the info bar at the bottom of the screen when the mouse cursor is positioned over the creature dwelling.

In original game:

Creature dwelling text info bar H2 01

In fheroes2:

Creature dwelling text info bar fh2 01

### PR DESCRIPTION
**Pull request by kant2002**
When hover over building display information about units which can be bought if this is that kind of building. For other building just display name of the building
Closes #1730

### CODE REVIEW

**Comment by kant2002** [comment]
Have following error.

```
../fheroes2/castle/castle.cpp:2804:13: error: declaration shadows a field of 'Castle' [-Werror,-Wshadow]
    Monster building( race, buildingId );
            ^
../fheroes2/castle/castle.h:242:9: note: previous declaration is here
    u32 building;
        ^
../fheroes2/castle/castle.cpp:2806:17: error: declaration shadows a field of 'Castle' [-Werror,-Wshadow]
        Monster dwelling( race, building.GetDwelling() );
                ^
../fheroes2/castle/castle.h:248:9: note: previous declaration is here
    u32 dwelling[CASTLEMAXMONSTER];

```
Isn't it a bit much? or can this be enforced on VC++ level too.

**Comment by ihhub** [comment]
Hi @kant2002 , none of warnings are enforced of developer's machine. We do this only on server's level. Sadly Visual Studio is not that flexible as gcc that's why right now we use -W4 flag and just don't fail the compilation.

**Comment by kant2002** [comment]
It was an annoyance to not be able fix that issue on dev machine

File: src/fheroes2/castle/castle.cpp [hunk 1]
```diff
@@ -2798,3 +2798,17 @@ std::string Castle::GetDescriptionBuilding( u32 build ) const
 
     return res;
 }
+
+std::string Castle::buildingStatusMessage( uint32_t buildingId ) const
+{
+    Monster buildingInfo( race, buildingId );
```
Comment (ihhub) [inline src/fheroes2/castle/castle.cpp:2804]: Please make this variable `const`.

File: src/fheroes2/castle/castle.cpp [hunk 2]
```diff
@@ -2798,3 +2798,17 @@ std::string Castle::GetDescriptionBuilding( u32 build ) const
...
+    if ( buildingInfo.GetDwelling() ) {
```
Comment (ihhub) [inline src/fheroes2/castle/castle.cpp:2805]: Please cache this variable as we call it twice `const uint32_t dwellingType = buildingInfo.GetDwelling();`.

File: src/fheroes2/castle/castle.cpp [hunk 3]
```diff
@@ -2798,3 +2798,17 @@ std::string Castle::GetDescriptionBuilding( u32 build ) const
 
     return res;
 }
+
+std::string Castle::buildingStatusMessage( uint32_t buildingId ) const
```
Comment (ihhub) [inline src/fheroes2/castle/castle.cpp:2802]: `buildingId` is a dwellingId for monsters. Why don't we try to do this `( buildingId & DWELLING_MONSTERS ) != 0` meaning that this is a monster dwelling? So the code would be:
```
if ( ( buildingId & DWELLING_MONSTERS ) == 0 ) {
    return GetStringBuilding( buildingId );
}
...
// the rest of the code.
```

File: src/fheroes2/castle/castle.cpp (hunk 1 above)
Comment (ihhub) [reply src/fheroes2/castle/castle.cpp:2804]: This variable might not needed based on the above comment.

File: src/fheroes2/castle/castle.cpp (hunk 2 above)
Comment (ihhub) [reply src/fheroes2/castle/castle.cpp:2805]: This variable might not needed based on the above comment.

File: src/fheroes2/castle/castle.cpp [hunk 4]
```diff
@@ -2798,3 +2798,17 @@ std::string Castle::GetDescriptionBuilding( u32 build ) const
...
+{
+    Monster buildingInfo( race, buildingId );
+    if ( buildingInfo.GetDwelling() ) {
+        Monster dwellingInfo( race, buildingInfo.GetDwelling() );
```
Comment (ihhub) [inline src/fheroes2/castle/castle.cpp:2806]: Please make this variable `const` and name it as `monster` as it is a monster :)

File: src/fheroes2/castle/castle.cpp [hunk 5]
```diff
@@ -2798,3 +2798,17 @@ std::string Castle::GetDescriptionBuilding( u32 build ) const
...
+        std::string msg_status = _( "Recruit %{name}" );
```
Comment (ihhub) [inline src/fheroes2/castle/castle.cpp:2807]: Please use camelCase variable naming: `std::string msgStatus( _( "Recruit %{name}" )  );`.

File: src/fheroes2/castle/castle.h [hunk 6]
```diff
@@ -193,6 +193,7 @@ class Castle : public MapPosition, public BitModes, public ColorBase, public Con
 
     std::string GetStringBuilding( u32 ) const;
     std::string GetDescriptionBuilding( u32 ) const;
+    std::string buildingStatusMessage( uint32_t ) const;
```
Comment (ihhub) [inline src/fheroes2/castle/castle.h:196]: Please add the name of input argument (old code was written in such weird way) and make the input parameter const as we don't modify it.

**Review by ihhub — CHANGES_REQUESTED** [review]
Hi @kant2002 , I left multiple comments here. Could you please take a look?

**Comment by kant2002** [comment]
@ihhub I fix all suggestions except `buildingId & DWELLING_MONSTERS ) == 0` change which seems to low-level and less intuitive for the reader. I hope and put trust in the compiler to revert conditions etc. Let me know if you have any concerns about this style

**Comment by ihhub** [comment]
> @ihhub I fix all suggestions except `buildingId & DWELLING_MONSTERS ) == 0` change which seems to low-level and less intuitive for the reader. I hope and put trust in the compiler to revert conditions etc. Let me know if you have any concerns about this style

Hi @kant2002 , it is not a low level code. This code:
```
    const Monster buildingInfo( race, buildingId );
    const uint32_t dwellingMonsterId = buildingInfo.GetDwelling();
    const bool isDwelling = dwellingMonsterId != 0;
```
essentially creates an instance of a class, we check dwelling id based on this instance and remove the object. My proposed solution does the check without extra memory usage. If you aren't sure about understanding by other developers let's put a comment describing why we do all this.

**Comment by kant2002** [comment]
Isn't the Monster class just glorified int?

**Comment by ihhub** [comment]
> Isn't the Monster class just glorified int?

Yes but we don't need to even create an instance of this class to do the job. Also there's an extra logic when we create it and call the function (invoking call stack usage).

**Comment by kant2002** [comment]
Also I do not fully understand what you trying to say about relationship between `buildingId` and `DWELLING_MONSTERS`. I still do not get how can I get `monsterId` from building id.

File: src/fheroes2/castle/castle.cpp [hunk 7]
```diff
@@ -2798,3 +2798,18 @@ std::string Castle::GetDescriptionBuilding( u32 build ) const
 
     return res;
 }
+
+std::string Castle::buildingStatusMessage( const uint32_t buildingId ) const
```
Comment (ihhub) [inline src/fheroes2/castle/castle.cpp:2802]: This is what I'm proposing:
```
std::string Castle::buildingStatusMessage( const uint32_t buildingId ) const
{
    // Check if building is a monster dwelling or its upgraded version
    if ( ( buildingId & DWELLING_MONSTERS ) == 0 && ( buildingId & DWELLING_UPGRADES ) == 0 ) {
        return GetStringBuilding( buildingId );
    }

    const Monster monster( race, buildingId );
    std::string msgStatus = _( "Recruit %{name}" );
    StringReplace( msgStatus, "%{name}", monster.GetMultiName() );
    return msgStatus;
}
```

Check that I changed GetName to GetMultiName as logically a dwelling contains more than 1 monster.

Comment (kant2002) [reply src/fheroes2/castle/castle.cpp:2802]: Ok. now it make sense. Regarding singular vs plural form. Testers are want pixel perfect match sometimes, so I do not move from original issues description. Now it make more sense.

Comment (ihhub) [reply src/fheroes2/castle/castle.cpp:2802]: I think everyone will agree with plural form. @Branikolog what do you think?

Comment (ihhub) [reply src/fheroes2/castle/castle.cpp:2802]: @LeHerosInconnu (sorry, missed you to mentioned in previous message), what do you think?

Comment (Branikolog) [reply src/fheroes2/castle/castle.cpp:2802]: Hi, @kant2002 and @ihhub !
Plural form of creatures is already implemented in fheroes2 recruit windows:
image
So in bottom textline should be the same should be.

Comment (LeHerosInconnu) [reply src/fheroes2/castle/castle.cpp:2802]: Taking into account the singular and plural improves the original game, so it's a good thing. :)

Comment (ihhub) [reply src/fheroes2/castle/castle.cpp:2802]: Great! @kant2002 , could you please incorporate these changes with plural form of monsters?

**Comment by ihhub** [comment]
@kant2002 thank you so much for fixing this issue!

**Pull request merged**